Can't build native image with GraalVM 22.3.0 Java 17 CE

A native image built via `mvn -Pnative -DskipTests package` with GraalVM 22.3.0 Java 17 CE fails at runtime with:

```
MyClass.java: error: com.sun.tools.javac.parser.Tokens$TokenKind
java.lang.NoClassDefFoundError: com.sun.tools.javac.parser.Tokens$TokenKind
	at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:350)
	at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:335)
	at com.google.googlejavaformat.java.JavaInput.<init>(JavaInput.java:277)
	at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:260)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:247)
	at com.google.googlejavaformat.java.FormatFileCallable.call(FormatFileCallable.java:45)
	at com.google.googlejavaformat.java.FormatFileCallable.call(FormatFileCallable.java:26)
	at java.base@17.0.5/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base@17.0.5/java.lang.Thread.run(Thread.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:775)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:203)
```

While this seems to be fixed by adding `-J--add-exports=...` `buildArgs` to the `native-maven-plugin` config, the produced native image still fails at runtime with a different error:

```
MyClass.java: error: com.google.googlejavaformat.java.java17.Java17InputAstVisitor. This exception was synthesized during native image building from a call to java.lang.Class.forName(String) with constant arguments.
java.lang.LinkageError: com.google.googlejavaformat.java.java17.Java17InputAstVisitor. This exception was synthesized during native image building from a call to java.lang.Class.forName(String) with constant arguments.
	at com.google.googlejavaformat.java.Formatter.format(Formatter.java:162)
	at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:271)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:247)
	at com.google.googlejavaformat.java.FormatFileCallable.call(FormatFileCallable.java:45)
	at com.google.googlejavaformat.java.FormatFileCallable.call(FormatFileCallable.java:26)
	at java.base@17.0.5/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base@17.0.5/java.lang.Thread.run(Thread.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:775)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:203)
Caused by: java.lang.ClassNotFoundException: com.google.googlejavaformat.java.java17.Java17InputAstVisitor. This exception was synthesized during native image building from a call to java.lang.Class.forName(String) with constant arguments.
	at com.google.googlejavaformat.java.Formatter.format(Formatter.java:157)
	... 10 more
```

I haven't found a fix for this yet.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -29,10 +29,10 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
-        java: [21, 17, 11]
+        java: [GraalVM, 21, 17, 11]
         experimental: [false]
         include:
-          # Only test on macos and windows with a single recent JDK to avoid a
+          # Only test on MacOS and Windows with a single recent JDK to avoid a
           # combinatorial explosion of test configurations.
           - os: macos-latest
             java: 21
@@ -43,6 +43,9 @@ jobs:
           - os: ubuntu-latest
             java: EA
             experimental: true
+          - os: ubuntu-latest
+            java: GraalVM
+            experimental: false
     runs-on: ${{ matrix.os }}
     continue-on-error: ${{ matrix.experimental }}
     steps:
@@ -58,12 +61,21 @@ jobs:
         with:
           website: jdk.java.net
           release: ${{ matrix.java }}
-      - name: "Set up JDK ${{ matrix.java }}"
-        if: ${{ matrix.java != 'EA' }}
+      - name: "Set up JDK ${{ matrix.java }} from Zulu"
+        if: ${{ matrix.java != 'EA' && matrix.java != 'GraalVM' }}
         uses: actions/setup-java@v4
         with:
           java-version: ${{ matrix.java }}
           distribution: "zulu"
+          cache: "maven"
+      - name: "Set up JDK ${{ matrix.java }}"
+        if: ${{ matrix.java == 'GraalVM' }}
+        uses: graalvm/setup-graalvm@v1
+        with:
+          java-version: "21"
+          distribution: "graalvm-community"
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          native-image-job-reports: "true"
           cache: "maven"
       - name: "Install"
         shell: bash
@@ -71,6 +83,9 @@ jobs:
       - name: "Test"
         shell: bash
         run: mvn test -B
+      - name: "Native"
+        if: ${{ matrix.java == 'GraalVM' }}
+        run: mvn -Pnative -DskipTests package -pl core -am && util/test-native.sh
 
   publish_snapshot:
     name: "Publish snapshot"

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -69,7 +69,7 @@ jobs:
         run: |
           git push origin "v${{ github.event.inputs.version }}"
 
-      - name: Add Jars to Release Entry
+      - name: Add Artifacts to Release Entry
         uses: softprops/action-gh-release@v0.1.14
         with:
           draft: true
@@ -77,5 +77,5 @@ jobs:
           tag_name: "v${{ github.event.inputs.version }}"
           target_commitish: ${{ env.TARGET_COMMITISH }}
           files: |
-            core/target/google-java-format-*.jar
+            core/target/google-java-format*
             eclipse_plugin/target/google-java-format-eclipse-plugin-*.jar

--- a/util/test-native.sh
+++ b/util/test-native.sh
@@ -1,0 +1,30 @@
+#!/usr/bin/env bash
+# Copyright 2024 The Google Java Format Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -euox pipefail
+
+time java -jar core/target/google-java-format-*-all-deps.jar || true
+
+status=-1
+chmod +x core/target/google-java-format
+if time core/target/google-java-format; then
+ status=0
+else
+ status=$?
+fi
+if [[ $status -ne 2 ]]; then
+  echo "google-java-format_linux (native) without arguments should have printed usage help and exited with 2, but did not :("
+  exit 1
+fi

